src/goto-symex/builtin_functions.cpp

```diff
@@ -748,277 +748,389 @@ void goto_symext::symex_va_arg(const expr2tc &lhs, const sideeffect2t &code)
+    auto name = to_union_type(type).member_names[selected_member_index];
+    auto member_type = to_union_type(type).members[selected_member_index];
+    member2tc member(member_type, src, name);
+
+    result->init_field = name;
+    result->datatype_members[0] =
+      gen_value_by_byte(member_type, member, value, num_of_bytes, offset);
+    return result;
   }
-  else
+
+  // Found a primitive! Just apply the function
+  return gen_byte_expression(type, src, value, num_of_bytes, offset);
+}
+
+void goto_symext::intrinsic_memset(
+  reachability_treet &art,
+  const code_function_call2t &func_call)
+{
+  /**
+     * @brief This function will try to initialize the object pointed by
+     * the address in a smarter way, minimizing the number of assignments.
+     * This is intend to optimize the behaviour of a memset operation:
+     * 
+     * memset(void* ptr, int value, size_t num_of_bytes)
+     * 
+     * - ptr can point to anything. We have to add checks!
+     * - value is interpreted as a uchar.
+     * - num_of_bytes must be known. If it is nondet, we will bump the call
+     * 
+     * In plain C, the objective of a call such as:
+     * 
+     * int a;
+     * memset(&a, value, num)
+     * 
+     * Would generate something as:
+     *      
+     * int temp = 0;
+     * for(int i = 0; i < num; i++) temp = byte | (temp << 8);
+     * a = temp;
+     * 
+     * This is just a simplification for understanding though. During the
+     * instrumentation size checks will be added, and also, the original
+     * bytes from `a` that were not overwritten must be mantained! 
+     * Arrays will need to be added up to an nth element.
+     * 
+     * In ESBMC though, we have 2 main methods of dealing with memory objects:
+     * 
+     * A. Heap objects, which are valid/invalid. They are the easiest to deal
+     *    with, as the dereference will actually return a big array of char to us.
+     *    For this case, we can just overwrite the members directly with the value
+     * 
+     * B. Stack objects, which are typed. It will be hard, this will require operations
+     *    which depends on the base type and also on padding.
+     * 
+     */
+
+  // 1. Check for the functions parameters and do the deref and processing!
+
+  assert(func_call.operands.size() == 3 && "Wrong memset signature");
+  auto &ex_state = art.get_cur_state();
+  if(ex_state.cur_state->guard.is_false())
+    return;
+
+  // 2. Try to generate the optimized object. If not, go to the default call
+  try
   {
-    bump_call();
+    // Get the argument
+    expr2tc arg0 = func_call.operands[0];
+    expr2tc arg1 = func_call.operands[1];
+    expr2tc arg2 = func_call.operands[2];
+
+    internal_deref_items.clear();
+    dereference2tc deref(get_empty_type(), arg0);
+    dereference(deref, dereferencet::INTERNAL);
+
+    // Is this pointing to a valid place?
+    if(!internal_deref_items.size())
+      throw std::invalid_argument("Couldn't deref the object");
```

Comment:
Can we remove the throws and simply guard against these error cases? Calling the code in the catch block when the if fails.

Reply:
replaced throws with an empty expression